#source ../env/bin/activate


rule all:
	shell: """
		snakemake -c1 -p ssRNA_fasta_dsDNA_fasta;
		snakemake -c1 -p ssRNA_id_dsDNA_fasta;
		snakemake -c1 -p ssRNA_fasta_dsDNA_MANE;
	"""

#Test with both ssRNA and dsDNA as fasta files - no species or dsDNA_predefined specified
rule ssRNA_fasta_dsDNA_fasta:
	input:
		ssRNA="test_data/ssRNA.fa",
		dsDNA="test_data/dsDNA_small.fa"
	shell: """
		python run_test.py --token {rule} --ssRNA_fasta_path {input.ssRNA} --dsDNA_fasta_or_bed_path {input.dsDNA} \
		--param_config test_data/3plex_default_params.yaml
	"""

#Test with ssRNA as transcript (compressed!) and dsDNA fasta- No species or dsDNA_predefined specified
rule ssRNA_id_dsDNA_fasta:
	input:
		ssRNA="test_data/ENST00000000233.fa.gz",
		dsDNA="test_data/dsDNA_small.fa"
	shell: """
		python run_test.py --token {rule} --ssRNA_fasta_path {input.ssRNA} --dsDNA_fasta_or_bed_path {input.dsDNA} \
		--param_config test_data/3plex_default_params.yaml
	"""

#Test with ssRNA fasta and dsDNA_predefined MANE - species = hsapiens
rule ssRNA_fasta_dsDNA_MANE:
	input:
		ssRNA="test_data/ssRNA.fa"
	shell: """
		python run_test.py --token {rule} --ssRNA_fasta_path {input.ssRNA} \
        --dsdna_target MANE.GRCh38.v1.1.refseq_genomic.Symbol.tss.1500_500 \
		--species "hsapiens" --param_config test_data/3plex_default_params.yaml
	"""

rule ssRNA_fasta_dsDNA_MANE_with_random:
	input:
		ssRNA="test_data/ssRNA.fa"
	shell: """
		python run_test.py --token {rule} --ssRNA_fasta_path {input.ssRNA} \
        --dsdna_target MANE.GRCh38.v1.1.refseq_genomic.Symbol.tss.1500_500 \
		--species "hsapiens" --param_config test_data/3plex_default_params.yaml
		--randomization 1
	"""